### PR TITLE
rgw: avoid infinite loop when deleting a bucket

### PR DESCRIPTION
**Pull request by Jeegn-Chen**
When deleting a bucket with an incomplete multipart upload that
has about 2000 parts uploaded, we noticed an infinite loop, which
stopped s3cmd from deleting the bucket forever.
Per check, when the bucket index was sharded (for example 128
shards), the original logic in
RGWRados::cls_bucket_list_unordered() did not calculate
the bucket shard ID correctly when the index key of a data
part was taken as the marker.

The issue is not necessarily reproduced each time. It will depend
on the key of the object. To reproduce it in 128-shard bucket,
we use 334 as the key for the incomplete multipart upload,
which will be located in Shard 127 (known by experiment). In this
setup, the original logic will usually come out a shard ID smaller
than 127 (since 127 is the largest one) from the marker and
thus a circle is constructed, which results in an infinite loop.

PS: Some times the bucket ID calculation may incorrectly going forward
instead of backward. Thus, the check logic may skip some shards,
which may have regular keys. In such scenarios, some non-empty buckets may
be deleted by accident.

Fixes: http://tracker.ceph.com/issues/49206
Signed-off-by: Jeegn Chen <jeegnchen@tencent.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by Jeegn-Chen** [comment]
jenkins retest this please

**Comment by Jeegn-Chen** [comment]
jenkins retest this please

File: src/rgw/rgw_rados.cc [hunk 1]
```diff
@@ -8579,6 +8579,14 @@ int RGWRados::cls_bucket_list_ordered(const DoutPrefixProvider *dpp,
   return 0;
 }
 
+static int parse_index_hash_source(const string& oid_wo_ns, string *index_hash_source) {
```
Comment (ivancich) [inline src/rgw/rgw_rados.cc:8582]: Please add a comment for the entire function that explains its narrow usage (removes everything from the second to last "." to the end) and that it's used to retrieve the hash source from an incomplete multipart entry.

Your use of namespace qualifiers is inconsistent. Please change `string` to `std::string`.

Your `if (...) return ...;`are against the ceph coding style (see _CodingStyle_ file at top level).

File: src/rgw/rgw_rados.cc [hunk 2]
```diff
@@ -8579,6 +8579,14 @@ int RGWRados::cls_bucket_list_ordered(const DoutPrefixProvider *dpp,
...
+  std::size_t found = oid_wo_ns.rfind('.');
+  if (found == std::string::npos || found < 1) return -EINVAL;
+  found = oid_wo_ns.rfind('.', found -1);
```
Comment (ivancich) [inline src/rgw/rgw_rados.cc:8585]: "-1" looks like negative one, where adding a space "- 1" clarifies that you're subtracting one.

File: src/rgw/rgw_rados.cc [hunk 3]
```diff
@@ -8647,7 +8648,19 @@ int RGWRados::cls_bucket_list_unordered(const DoutPrefixProvider *dpp,
     } else {
       // so now we have the key used to compute the bucket index shard
       // and can extract the specific shard from it
-      current_shard = svc.bi_rados->bucket_shard_index(obj_key.name, num_shards);
+      if (obj_key.ns == RGW_OBJ_NS_MULTIPART) {
+        // Use obj_key.ns == RGW_OBJ_NS_MULTIPART instead of
```
Comment (ivancich) [inline src/rgw/rgw_rados.cc:8652]: Thank you for this comment -- it clarifies a lot! ("implementation" is misspelled.)

File: src/rgw/rgw_rados.cc [hunk 4]
```diff
@@ -8647,7 +8648,19 @@ int RGWRados::cls_bucket_list_unordered(const DoutPrefixProvider *dpp,
...
+        // the implememtation relying on MultipartMetaFilter
+        // because MultipartMetaFilter only checks .meta suffix, which may
+        // exclude data multiparts but include some regular objects with .meta suffix
+        // by mistake.
+        string index_hash_source;
+        r = parse_index_hash_source(obj_key.name, &index_hash_source);
+        if (r < 0) return r;
```
Comment (ivancich) [inline src/rgw/rgw_rados.cc:8659]: Same as above w.r.t. coding style.

**Review by ivancich — APPROVED** [review]
I have some minor nits, but this is excellent! Thank you!

**Comment by Jeegn-Chen** [comment]
> I have some minor nits, but this is excellent! Thank you!

Thanks a lot for the review! I will address the comments soon.

**Comment by ivancich** [comment]
@harishmunjulur This was labeled as being tested by you about a month ago. Are you still testing this?

**Comment by Jeegn-Chen** [comment]
@ivancich @harishmunjulur 
Do we plan to make the commit merged or another preferred fix has been accepted?

**Comment by ivancich** [comment]
> @ivancich @harishmunjulur
> Do we plan to make the commit merged or another preferred fix has been accepted?

This pr stalled because it was labeled as being under testing by @harishmunjulur 3.5 months ago, so no one else picked it up for a QA run. And then it's unclear what happened with @harishmunjulur. Sorry!!

I'm going to remove @harishmunjulur's testing label so it can get picked up for QA again.

**Comment by harishmunjulur** [comment]
@Jeegn-Chen @ivancich 

https://trello.com/c/yVVuvuGb

http://pulpito.ceph.com/hmunjulu-2021-03-24_00:22:35-rgw-wip-hmunjulu01-testing-2021-03-23-1135-distro-basic-smithi/

**Comment by ivancich** [comment]
Thank you @harishmunjulur . I reviewed the teuthology errors and none seem applicable. Therefore I'll merge this PR. Thanks, everyone!

@Jeegn-Chen I don't know if you have time, but it would be great to merge another PR with a test that demonstrates the failure of listing to advance correctly without this fix. Thanks for considering.

**Pull request merged**